src/github/mod.rs

```diff
@@ -0,0 +1,220 @@
+#![allow(unused_must_use)]
+use crate::{
+    embeddings::{Embeddings, EmbeddingsModel},
+    prelude::*,
+};
+use rayon::prelude::{IntoParallelIterator, ParallelIterator};
+use serde::{Deserialize, Serialize};
+use std::io::Read;
+
+#[derive(Debug, Default, Serialize)]
+pub struct File {
+    pub path: String,
+    pub content: String,
+    pub length: usize,
+}
+
+impl ToString for File {
+    fn to_string(&self) -> String {
+        format!(
+            "File path: {}\nFile length: {} bytes\nFile content: {}",
+            &self.path, &self.length, &self.content
+        )
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct FileEmbeddings {
+    pub path: String,
+    pub embeddings: Embeddings,
+}
+
+#[derive(Debug)]
+pub struct RepositoryEmbeddings {
+    pub repo_id: String,
+    pub file_embeddings: Vec<FileEmbeddings>,
+}
+
+#[derive(Serialize, Debug)]
+pub struct RepositoryFilePaths {
+    pub repo_id: String,
+    pub file_paths: Vec<String>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct Repository {
+    pub owner: String,
+    pub name: String,
+    pub branch: String,
+}
+
+impl ToString for Repository {
+    fn to_string(&self) -> String {
+        format!("{}-{}-{}", &self.owner, &self.name, &self.branch)
+    }
+}
+
+pub async fn embed_repo<M: EmbeddingsModel + Send + Sync>(
+    repository: &Repository,
+    files: Vec<File>,
+    model: &M,
+) -> Result<RepositoryEmbeddings> {
+    let file_embeddings: Vec<FileEmbeddings> = files
+        .into_par_iter()
+        .filter_map(|file| {
+            let embed_content = file.to_string();
+            let embeddings = model.embed(&embed_content).unwrap();
+            Some(FileEmbeddings {
+                path: file.path,
+                embeddings,
+            })
+        })
+        .collect();
+
+    Ok(RepositoryEmbeddings {
+        repo_id: repository.to_string(),
+        file_embeddings,
+    })
+}
+
+pub async fn fetch_repo_files(repository: &Repository) -> Result<Vec<File>> {
+    let Repository {
+        owner,
+        name,
+        branch,
+    } = repository;
+
+    let url = format!("https://github.com/{owner}/{name}/archive/{branch}.zip");
```

Comment:
It'd be interesting one day to extend this beyond zip archieves on github and make this agnostic to _any_ git repository out on the web. But, for now, this is great.